src/Juvix/Compiler/Core/Transformation/UnrollRecursion.hs

```diff
@@ -1,7 +1,88 @@
 module Juvix.Compiler.Core.Transformation.UnrollRecursion (unrollRecursion) where
 
+import Data.HashMap.Strict qualified as HashMap
+import Juvix.Compiler.Core.Data.IdentDependencyInfo
+import Juvix.Compiler.Core.Data.InfoTableBuilder
+import Juvix.Compiler.Core.Extra
 import Juvix.Compiler.Core.Transformation.Base
 
--- TODO: not yet implemented / at first only check for recursion and give an error
 unrollRecursion :: InfoTable -> InfoTable
-unrollRecursion tab = tab
+unrollRecursion tab =
+  let (mp, (tab', ())) =
+        run $
+          runState @(HashMap Symbol Symbol) mempty $
+            runInfoTableBuilder tab $
+              forM_ (buildSCCs (createIdentDependencyInfo tab)) goSCC
+   in mapIdentSymbols mp $ pruneInfoTable tab'
+  where
+    mapIdentSymbols :: HashMap Symbol Symbol -> InfoTable -> InfoTable
+    mapIdentSymbols mp = over infoMain adjustMain . mapAllNodes (umap go)
+      where
+        go :: Node -> Node
+        go = \case
+          NIdt idt@Ident {..} ->
+            case HashMap.lookup _identSymbol mp of
+              Just sym' -> NIdt idt {_identSymbol = sym'}
+              Nothing -> NIdt idt
+          node -> node
+
+        adjustMain :: Maybe Symbol -> Maybe Symbol
+        adjustMain = \case
```

Comment:
use `<$>` instead of `\case`

Reply:
I changed it to use `fmap` and a lambda. I don't know how you want to use `<$>` here.